corehq/apps/hqadmin/couch_domain_utils.py

```diff
@@ -0,0 +1,148 @@
+
+from corehq.pillows.xform import get_xform_pillow
+from corehq.util.couch import bulk_get_revs
+from corehq.apps.hqcase.management.commands.backfill_couch_forms_and_cases import (
+    create_form_change_meta,
+)
+
+
+def cleanup_stale_es_on_couch_domains(
+    start_date=None, end_date=None, domains=None, stdout=None
+):
+    """
+    This is the response to https://dimagi-dev.atlassian.net/browse/SAAS-11489
+    and basically makes sure that there are no stale docs in the most active
+    domains still using the couch db backend until we can get them migrated.
+    """
+    end = end_date or datetime.datetime.utcnow()
+    start = start_date or (end - datetime.timedelta(days=5))
+
+    couch_domains = domains or ACTIVE_COUCH_DOMAINS.get_enabled_domains()
+
+    for domain in couch_domains:
+        form_ids, has_discrepancies = _get_all_form_ids(domain, start, end)
+        if has_discrepancies:
+            metrics_gauge(
+                'commcare.es.couch_domain.couch_discrepancy_detected',
+                1,
+                tags={
+                    'domain': domain,
+                }
+            )
+            if stdout:
+                stdout.write(f"\nFound discrepancies in form counts for domain {domain}")
+        forms_not_in_es = _get_forms_not_in_es(form_ids)
+        if forms_not_in_es:
+            metrics_gauge(
+                'commcare.es.couch_domain.stale_docs_in_es',
+                len(forms_not_in_es),
+                tags={
+                    'domain': domain,
+                }
+            )
+            if stdout:
+                stdout.write(f"\nFound {len(forms_not_in_es)} forms not in es "
+                             f"for {domain}")
+            changes = _get_changes(domain, forms_not_in_es)
+            form_es_processor = get_xform_pillow().processors[0]
+            for change in changes:
+                form_es_processor.process_change(change)
+
+
+def _get_couch_form_ids(domain, start, end):
+    "Get form IDs from the 'by_domain_doc_type_date/view' couch view"
+    form_ids = set()
+    for doc_type in ['XFormInstance']:
+        startkey = [domain, doc_type, json_format_datetime(start)]
+        endkey = [domain, doc_type, json_format_datetime(end)]
+        results = XFormInstance.get_db().view(
+            'by_domain_doc_type_date/view', startkey=startkey, endkey=endkey,
+            reduce=False, include_docs=False
+        )
+        form_ids.update({row['id'] for row in results})
+    return form_ids
+
+
+def _get_all_form_ids(domain, start, end):
+    """Get form IDs from couch multiple times to ensure we get them all
+    This should not be necessary but for some reason we are seeing partial results.
+    Also returns a boolean in the second parameter to indicate if there were
+    any discrepancies in the results.
+    """
+    all_form_ids = None
+    previous_form_ids = None
+    form_id_differences = []
+    for i in range(10):
```

Comment:
Where did the magic number 10 come from? How do we know it will always get all ids with 10 queries?

Is there any way to query each couch node independently and then merge the results? That would require only 3 queries. I'm assuming the results returned by a given node are always the same, but do we know if this is true?

Reply:
Maybe we could use a [quorum parameter](https://docs.couchdb.org/en/stable/cluster/sharding.html#quorum) like `?r=3`? Not sure if couch will combine the results from all shards? Is it smart enough to detect its own internal inconsistencies?

Reply:
i'll look into this tomorrow, but for now, I just want to get this fix in to bring this down to a P2 and I've been waiting a while for this build to pass.

Reply:
10 came originally from simon's scripts

Reply:
Would it be possible to query each node directly?

Reply:
> 10 came originally from simon's scripts. it generally gets the job done

I'm uncomfortable with this. Simon's script was made to investigate a potential problem, not as a proven solution to that problem. I don't know what the odds are of running 10 queries and not hitting every node, but I imagine it is affected by cluster load and possibly other factors as well. It seems possible that this is not 100% reliable.